client/lib/protect-form/index.jsx

```diff
@@ -57,6 +67,41 @@ export const protectForm = WrappedComponent => {
 	};
 };
 
+/*
+ * Declarative variant that takes a 'isChanged' prop.
+ */
+export class ProtectForm extends Component {
+	static propTypes = {
+		isChanged: PropTypes.bool.isRequired
+	}
+
+	// This function binding makes the listener attached to window unique for each
+	// component instance.
+	warnIfChanged = warnIfChanged.bind();
+
+	componentDidMount() {
+		window.addEventListener( 'beforeunload', this.warnIfChanged );
+		if ( this.props.isChanged ) {
+			markChanged( this );
+		}
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'beforeunload', this.warnIfChanged );
+		markSaved( this );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.isChanged !== this.props.isChanged ) {
+			nextProps.isChanged ? markChanged( this ) : markSaved( this );
+		}
+	}
+
+	render() {
+		return null;
+	}
+}
```

Comment:
It doesn't have any children -- it's just a guard component that does stuff on mount/unmount/prop change, and renders nothing. Very similar to, e.g., `<QueryGeo />`.